artiq/dashboard/moninj.py

```diff
@@ -395,25 +402,47 @@ async def core_connector(self):
             if self.core_connection is not None:
                 await self.core_connection.close()
                 self.core_connection = None
-            new_core_connection = CommMonInj(self.monitor_cb, self.injection_status_cb,
-                    self.disconnect_cb)
+
+            async def connect(strat: CoreDeviceResolveStrategy) -> bool:
+                """
+                Run connection to the core device using different strategies
+                :param strat: the connection strategy to use
+                :return True/False if a connection is established with that strategy
+                """
+
+                new_core_connection = CommMonInj(self.monitor_cb, self.injection_status_cb,
+                                                 self.disconnect_cb)
+                try:
+                    if strat == CoreDeviceResolveStrategy.DIRECT:
+                        await new_core_connection.connect(self.core_addr, 1383)
+                    elif strat == CoreDeviceResolveStrategy.MASTER_PROXY:
+                        await new_core_connection.connect(self.core_master_fallback[0], self.core_master_fallback[1])
+                except asyncio.CancelledError:
+                    raise
+                except:
+                    logger.error("failed to connect to core device moninj", exc_info=True)
+                    await asyncio.sleep(10.)
+                    self.reconnect_core.set()
+                    return False
+                else:
+                    self.core_connection = new_core_connection
+                    for ttl_channel in self.ttl_widgets.keys():
+                        self.setup_ttl_monitoring(True, ttl_channel)
+                    for bus_channel, channel in self.dds_widgets.keys():
+                        self.setup_dds_monitoring(True, bus_channel, channel)
+                    for spi_channel, channel in self.dac_widgets.keys():
+                        self.setup_dac_monitoring(True, spi_channel, channel)
+                return True
+
+            # a very rudimentary core device resolve strategy
+            # if the dashboard cannot connect to the core device directly, we use the fallback provided by master
+            # TODO: extend this to a retry counter based approach
```

Comment:
With a firewall that drops packets, it can take dozens of seconds or more before the connection fails, with a corresponding and unacceptable delay before moninj works. Anyway this is not necessary, just use the proxy all the time.